Sources/Machine/py_abstract_machine.cc

```diff
@@ -0,0 +1,215 @@
+                   // compiler that this branch never returns (MPI_Abort is not
+                   // marked [[noreturn]]).
+  } catch (...) {
+    if (ShouldIDoIO()) {
+      std::fprintf(
+          stderr,
+          "[NetKet] Fatal error: exception was thrown in a `noexcept` context\n"
+          "[NetKet] Aborting...\n");
+    }
+    MPI_Abort(MPI_COMM_WORLD, -1);
+    std::abort();  // This call is unreachable and is here just to tell the
+                   // compiler that this branch never returns (MPI_Abort is not
+                   // marked [[noreturn]]).
+  }
+}
+
+int PyAbstractMachine::Npar() const {
+  PYBIND11_OVERLOAD_PURE_NAME(
+      int,                  /* Return type */
+      AbstractMachine,      /* Parent class */
+      "_number_parameters", /* Name of the function in Python */
+      Npar,                 /* Name of function in C++ */
+  );
+}
+
+int PyAbstractMachine::Nvisible() const {
+  PYBIND11_OVERLOAD_PURE_NAME(
+      int,               /* Return type */
+      AbstractMachine,   /* Parent class */
+      "_number_visible", /* Name of the function in Python */
+      Nvisible,          /* Name of function in C++ */
+  );
+}
+
+bool PyAbstractMachine::IsHolomorphic() const noexcept {
+  return ShouldNotThrow([this]() {
+    PYBIND11_OVERLOAD_PURE_NAME(
+        bool,              /* Return type */
+        AbstractMachine,   /* Parent class */
+        "_is_holomorphic", /* Name of the function in Python */
+        IsHolomorphic,     /* Name of function in C++ */
+    );
+  });
+}
+
+PyAbstractMachine::VectorType PyAbstractMachine::GetParameters() {
+  PYBIND11_OVERLOAD_PURE_NAME(
+      VectorType,        /* Return type */
+      AbstractMachine,   /* Parent class */
+      "_get_parameters", /* Name of the function in Python */
+      GetParameters,     /* Name of function in C++ */
+  );
+}
+
+void PyAbstractMachine::SetParameters(VectorConstRefType pars) {
+  PYBIND11_OVERLOAD_PURE_NAME(
+      void,              /* Return type */
+      AbstractMachine,   /* Parent class */
+      "_set_parameters", /* Name of the function in Python */
+      SetParameters,     /* Name of function in C++ */
+      pars);
+}
+
+void PyAbstractMachine::InitRandomPars(int const seed, double const sigma) {
+  VectorType par(Npar());
+  netket::RandomGaussian(par, seed, sigma);
+  SetParameters(par);
+}
+
+Complex PyAbstractMachine::LogVal(VisibleConstType v) {
+  PYBIND11_OVERLOAD_PURE_NAME(Complex,         /* Return type */
+                              AbstractMachine, /* Parent class */
+                              "log_val", /* Name of the function in Python */
+                              LogVal,    /* Name of function in C++ */
+                              v);
+}
+
+Complex PyAbstractMachine::LogVal(VisibleConstType v,
+                                  const LookupType & /*unused*/) {
+  return LogVal(v);
+}
+
+void PyAbstractMachine::InitLookup(VisibleConstType /*unused*/,
```

Comment:
Is this unused because we cannot derive a machine in Python using the look-up tables? Why we cannot do that?

Reply:
Like discussed, with PyTorch and deeper architectures it makes little sense to implement look-up tables.

Reply:
True, but there is other wave-functions for which it still makes sense. If it's not too much of an effort, I believe we should support also look-up tables. Is there some technical problem there?

Reply:
As for technical reasons... converting `std::vector` to `numpy.ndarray` and back does a lot of unnecessary allocations. These allocations defeat the purpose of having look-up tables as an optimisation in the first place.

So to actually benefit from look-up tables, we need to rethink the way they're used on the C++ side. This is a pretty major refactoring, I'm afraid, and as such is a topic for a different PR/issue.

Reply:
This falls within issue #52 , I agree, but it's also important to do at some point in the future